src/storage/wal_replay.cpp

```diff
@@ -763,58 +764,96 @@ void WriteAheadLogDeserializer::ReplayRowGroupData() {
 	// delete the tuples from the current table
 	TableDeleteState delete_state;
-	for (idx_t i = 0; i < chunk.size(); i++) {
+	for (idx_t i = 0; i < info->chunk->size(); i++) {
 		row_ids[0] = source_ids[i];
 		state.current_table->GetStorage().Delete(delete_state, context, row_identifiers, 1);
 	}
+	return nullptr;
 }
 
-void WriteAheadLogDeserializer::ReplayUpdate() {
-	auto column_path = deserializer.ReadProperty<vector<column_t>>(101, "column_indexes");
-
-	DataChunk chunk;
-	deserializer.ReadObject(102, "chunk", [&](Deserializer &object) { chunk.Deserialize(object); });
+unique_ptr<TableDataUpdateInfo> WriteAheadLogDeserializer::ReplayUpdate() {
+	auto info = make_uniq<TableDataUpdateInfo>();
+	info->column_indexes = deserializer.ReadProperty<vector<column_t>>(101, "column_indexes");
+	info->chunk = make_uniq<DataChunk>();
+	deserializer.ReadObject(102, "chunk", [&](Deserializer &object) { info->chunk->Deserialize(object); });
 
 	if (DeserializeOnly()) {
-		return;
+		return info;
 	}
 	if (!state.current_table) {
 		throw InternalException("Corrupt WAL: update without table");
 	}
 
-	if (column_path[0] >= state.current_table->GetColumns().PhysicalColumnCount()) {
+	if (info->column_indexes[0] >= state.current_table->GetColumns().PhysicalColumnCount()) {
 		throw InternalException("Corrupt WAL: column index for update out of bounds");
 	}
 
 	// remove the row id vector from the chunk
-	auto row_ids = std::move(chunk.data.back());
-	chunk.data.pop_back();
+	auto row_ids = std::move(info->chunk->data.back());
+	info->chunk->data.pop_back();
 
 	// now perform the update
-	state.current_table->GetStorage().UpdateColumn(*state.current_table, context, row_ids, column_path, chunk);
+	state.current_table->GetStorage().UpdateColumn(*state.current_table, context, row_ids, info->column_indexes,
+	                                               *info->chunk);
+	return nullptr;
 }
 
-void WriteAheadLogDeserializer::ReplayCheckpoint() {
+unique_ptr<ParseInfo> WriteAheadLogDeserializer::ReplayCheckpoint() {
 	state.checkpoint_id = deserializer.ReadProperty<MetaBlockPointer>(101, "meta_block");
+	auto info = make_uniq<ParseInfo>(ParseInfoType::CHECKPOINT_INFO);
+	if (DeserializeOnly()) {
+		return info;
+	}
+	return nullptr;
+}
+
+WALReader::WALReader(AttachedDatabase &database, unique_ptr<BufferedFileReader> reader_p) {
+	Connection con(database.GetDatabase());
+	reader = std::move(reader_p);
+	if (reader->Finished()) {
+		return;
+	}
+	// WAL file version, so Open() has the reader positioned at the expected offset.
+	//
+	// Simplified logic from ReplayVersion(). Note that we can't call the deserializer
+	// directly here because of chicken and egg.
+	//
+	// An alternative is to call reader->Read<idx_t>(), ignore the return value and use
+	// duckdb::WAL_VERSION_NUMBER to set the wal_version.
+	auto wal_version = (reader->Read<idx_t>() >> 40) & 0xFF;
+	state = make_uniq<ReplayState>(database, *con.context);
+	state->wal_version = wal_version;
 }
```

Comment:
I suggesting changing the code to the following, which then reads the first entry which fetches the correct WAL version and updates the state. 

```suggestion
    deserializer = WriteAheadLogDeserializer::Open(*state, *reader, true);
    _ = deserializer.DeserializeEntry();
}
```

Reply:
```
	reader->Read<idx_t>();
	state = make_uniq<ReplayState>(database, *con.context);
	state->wal_version = 2;  // There is a named constant in another file, will need to move it to a header
```

is the only alternative that passes unit tests.